Implement unrectifyImage()

mintar opened the pull request:
This PR removes the "PinholeCameraModel::unrectifyImage is unimplemented" exception and actually implements the function! :)

Closes #355.

I have also added a unit test. This PR works just fine when no ROI is set in the camera_info message. However, I am unsure how to handle the ROI here. As far as I can tell, since we are doing unrectification, i.e., processing an already rectified image, the ROI is not relevant here.

**Question to the maintainers:** Do we need to handle the ROI for unrectifying images?

* If **yes**, I'll have to address the part marked `// TODO: use K_binned, P_binned, binned_resolution`. I have copied most of the ROI handling code over from `initRectificationMaps`; only the part where these three variables are actually used has to be fixed.
* If **no**, I'll remove all the unnecessary ROI handling code and simplify the PR.

mintar committed:
Implement unrectifyImage()

Closes #355.

mintar commented:
Closing and reopening to re-trigger build farm tests.

mintar commented:
@mjcarroll : Any opinions on whether we have to handle the ROI in unrectifyImage? (see my question in the original description)

mjcarroll commented:
> Question to the maintainers: Do we need to handle the ROI for unrectifying images?

I did a re-read of [REP 104](https://www.ros.org/reps/rep-0104.html) and couldn't justify a strong argument either way.

I'm trying to think of a case where it would be relevant, since as you pointed out, we're already dealing with rectified images.
I suppose you could have a rectified image and contains entirely valid pixels, with an ROI indicating that some of the original raw pixels were discarded.  In that case, if you wanted to project the rectified ROI pixels back in the correct locations in the full resolution, you would have to take the ROI into account.

Does this seem like a valid use case?

mjcarroll commented:
Alternatively, a datapoint in support of removing would be that we don't deal with the ROI in `unrectifyPoint` currently, either: https://github.com/ros-perception/vision_opencv/blob/noetic/image_geometry/src/pinhole_camera_model.cpp#L377

mintar commented:
Ok, in that case, I would prefer removing the half-baked ROI support from this PR. That way, `unrectifyImage()` would be consistent with `unrectifyPoint()` (and nobody seems to have missed It there). Would that be okay for you?

mjcarroll commented:
> Ok, in that case, I would prefer removing the half-baked ROI support from this PR. That way, unrectifyImage() would be consistent with unrectifyPoint() (and nobody seems to have missed It there). Would that be okay for you?

Good with me.

mintar committed:
initRectificationMaps: Handle empty roi size correctly

According to REP 104, roi.height (resp. roi.width) = 0 should be treated
the same as full resolution.

mintar committed:
Add K and P args to [un]rectifyPoint, projectPixelTo3dRay

This allows using different matrices than K_ and P_ (which already
include binning and ROI), and is required for unrectifyImage.

mintar commented:
So, I've dug into the code again, and found out that `unrectifyPoint` **does** actually deal with ROI and binning correctly, because `K_` and `P_` already include binning and ROI. So I decided to make `unrectifyImage` also handle ROI and binning correctly. It took me longer than I'd like to admit, but I'm sure that now `unrectifyImage` works as expected, and I have the tests to prove it!

@mjcarroll : This PR is now ready for review.

mintar commented:
Thanks for the review!
